src/VisualStudio/Core/Def/ProjectSystem/VisualStudioProjectFactory.cs

```diff
@@ -45,40 +53,65 @@ public VisualStudioProjectFactory(
         [ImportMany] IEnumerable<Lazy<IDynamicFileInfoProvider, FileExtensionsMetadata>> fileInfoProviders,
         IVisualStudioDiagnosticAnalyzerProviderFactory vsixAnalyzerProviderFactory,
         [ImportMany] IEnumerable<IAnalyzerAssemblyRedirector> analyzerAssemblyRedirectors,
-        IVsService<SVsSolution, IVsSolution2> solution2)
+        IVsService<SVsSolution, IVsSolution2> solution2,
+        IAsynchronousOperationListenerProvider listenerProvider)
     {
         _threadingContext = threadingContext;
         _visualStudioWorkspaceImpl = visualStudioWorkspaceImpl;
         _dynamicFileInfoProviders = fileInfoProviders.AsImmutableOrEmpty();
         _vsixAnalyzerProviderFactory = vsixAnalyzerProviderFactory;
         _analyzerAssemblyRedirectors = analyzerAssemblyRedirectors.AsImmutableOrEmpty();
         _solution2 = solution2;
-    }
 
-    public Task<ProjectSystemProject> CreateAndAddToWorkspaceAsync(string projectSystemName, string language, CancellationToken cancellationToken)
-        => CreateAndAddToWorkspaceAsync(projectSystemName, language, new VisualStudioProjectCreationInfo(), cancellationToken);
+        _uiContextUpdateWorkQueue = new AsyncBatchingWorkQueue<string>(
+            DelayTimeSpan.NearImmediate,
+            UpdateUIContextsAsync,
+            StringComparer.OrdinalIgnoreCase,
+            listenerProvider.GetListener(FeatureAttribute.Workspace),
+            _threadingContext.DisposalToken);
+
+        _initializationTask = _threadingContext.JoinableTaskFactory.RunAsync(
+            async () =>
+            {
+                var cancellationToken = _threadingContext.DisposalToken;
+                // HACK: Fetch this service to ensure it's still created on the UI thread; once this is
+                // moved off we'll need to fix up it's constructor to be free-threaded.
+
+                await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(alwaysYield: true, cancellationToken);
+                _visualStudioWorkspaceImpl.Services.GetRequiredService<VisualStudioMetadataReferenceManager>();
+
+                _visualStudioWorkspaceImpl.SubscribeExternalErrorDiagnosticUpdateSourceToSolutionBuildEvents();
+                _visualStudioWorkspaceImpl.SubscribeToSourceGeneratorImpactingEvents();
+
+                // Since we're on the UI thread here anyways, use that as an opportunity to grab the
+                // IVsSolution object and solution file path.
+                var solution = await _solution2.GetValueOrNullAsync(cancellationToken).ConfigureAwait(true);
+                _solutionFilePath = solution != null && ErrorHandler.Succeeded(solution.GetSolutionInfo(out _, out var filePath, out _))
+                    ? filePath
+                    : null;
```

Comment:
This will only run once, which will break things badly if the user closes one solution and opens another. That said though, in the last few releases a bunch of the solution-loading mechanics in the shell were rewritten to be managed code, so maybe there's a free-threaded way to get this now. Or this could subscribe to an event and update it whenever that event fires.

Reply:
I'm curious if this worked before, I'll test it out.

Reply:
The GetSolutionInfo call would differ after the solution changed. Need to handle that better.

Reply:
IVsBackgroundSolution seems to work with the project types I've tested which allows obtaining the solution path off the main thread.